### PR TITLE
The `length` validator only takes effect for parameters with types that support `#length` method

### PR DESCRIPTION
**Pull request by OuYangJinTing**
Thanks to @dhruvCW for adding the length validator(#2437).

I found that in the case of optional parameters, passing in nil will lead to an exception. This behavior is not very friendly.
For nil, it should be a better choice to skip verification directly.

### CODE REVIEW

File: lib/grape/validations/validators/length_validator.rb
Comment (dhruvCW) [inline lib/grape/validations/validators/length_validator.rb:21]: shouldn't this only return if the `required` field is false. We need to store that in an instance member variable now as well.

**Comment by dhruvCW** [comment]
>I found that in the case of optional parameters, passing in nil will lead to an exception. This behavior is not very friendly. For nil, it should be a better choice to skip verification directly.

Ah nice catch 👏  for some reason I thought the validator won't be called if the parameter was not present (and optional) 😅

**Comment by dblock** [comment]
@OuYangJinTing care to answer the q above and update CHANGELOG, please?

File: lib/grape/validations/validators/length_validator.rb
Comment (OuYangJinTing) [reply lib/grape/validations/validators/length_validator.rb:21]: Sorry, I don't understand that return here will lead to problems such as instance variables. If possible, could you please explain it more clearly?

Comment (dhruvCW) [reply lib/grape/validations/validators/length_validator.rb:21]: sorry for not being clear 😅 . I was wondering if we should be skipping the validation for required parameters or not ?

After thinking about it a `nil` required parameter would fail validation anyways so this should be fine ✅

Comment (OuYangJinTing) [reply lib/grape/validations/validators/length_validator.rb:21]: Thank～

**Comment by OuYangJinTing** [comment]
> Ah nice catch 👏 for some reason I thought the validator won't be called if the parameter was not present (and optional) 😅

If the optional parameters are only passed, the verification will also be performed.
https://github.com/ruby-grape/grape/blob/v2.1.2/lib/grape/validations/validators/base.rb#L53

**Comment by OuYangJinTing** [comment]
@dblock Done ～

File: spec/grape/validations/validators/length_spec.rb
Comment (dblock) [inline spec/grape/validations/validators/length_spec.rb:197]: Nit: does not raise an error.

File: README.md
Comment (dblock) [inline README.md:1716]: Quote both `nil`'s.

File: CHANGELOG.md
Comment (dblock) [inline CHANGELOG.md:5]: Let's say "Allow `nil` in `length` validator"?

**Review by dblock — COMMENTED** [review]
I have some nits, but more importantly, do we have a test for "If you do not allow nil, you can use the `allow_blank: false` option." If not please add?

**Comment by OuYangJinTing** [comment]
> I have some nits, but more importantly, do we have a test for "If you do not allow nil, you can use the `allow_blank: false` option." If not please add?

@dblock Thank you for your feedback. At present, all feedback has been resolved. Please take a look when you have time.

File: spec/grape/validations/validators/length_spec.rb
Comment (dblock) [inline spec/grape/validations/validators/length_spec.rb:221]: Nit: lowercase "When" please?

File: README.md
Comment (dblock) [inline README.md:1716]: Add "want to", so "If you do not want to allow..."

**Review by dblock — COMMENTED** [review]
Looks good, one more typo below.

File: README.md
Comment (dhruvCW) [reply README.md:1716]: um this will be problematic if you want to allow zero length parameters. For example 

```ruby
requires :some_list, type: [Integer], length: { min: 0, max: 10 }, allow_blank: false
requires :some_string, type: String, length: { min: 0 }, allow_blank: false
```

would both have validation errors if we provided the following parameters `{ some_list: [], some_string: '' }`

IMO this is incorrect behaviour.

File: lib/grape/validations/validators/length_validator.rb
Comment (dblock) [reply lib/grape/validations/validators/length_validator.rb:21]: A `nil` value is a perfectly valid value, that's why we have `allow_blank` (to distinguish the cases when the parameter itself is required and a value for the parameter is required).

File: README.md
Comment (dblock) [reply README.md:1716]: I think this is exactly the expected behavior. Why do you think this is incorrect @dhruvCW?

Comment (dhruvCW) [reply README.md:1716]: I see like I mentioned combining the two can lead to unexpected results since allow_blank validates both nil as well as empty values. I personally then would appreciate this specific case to be documented so that users are aware about this potential pitfall.

File: lib/grape/validations/validators/length_validator.rb
Comment (dhruvCW) [reply lib/grape/validations/validators/length_validator.rb:21]: I see I guess I contravened required with nil values.

File: README.md
Comment (dblock) [reply README.md:1716]: Yes, we should, @OuYangJinTing add specs for ^ and README pls?

File: lib/grape/validations/validators/length_validator.rb
Comment (dblock) [reply lib/grape/validations/validators/length_validator.rb:21]: I know it's confusing. We have gone back and forth on this 🙈

File: README.md
Comment (OuYangJinTing) [reply README.md:1716]: @dblock @dhruvCW Sorry, I'm a little busy with work these two days, and I didn't give timely feedback.
Based on this consideration, I think it should be better to add an allow_nil option to the `Grape::Validations::Validators::LengthValidator`. e.g:

```ruby
module Grape
  module Validations
    module Validators
      class LengthValidator < Base
        def initialize(attrs, options, required, scope, **opts)
          @min = options[:min]
          @max = options[:max]
          @allow_nil = options.key?(:allow_nil) ? options[:allow_nil] : true # allow `allow_nil` option, default to `true`

          super

          raise ArgumentError, 'min must be an integer greater than or equal to zero' if !@min.nil? && (!@min.is_a?(Integer) || @min.negative?)
          raise ArgumentError, 'max must be an integer greater than or equal to zero' if !@max.nil? && (!@max.is_a?(Integer) || @max.negative?)
          raise ArgumentError, "min #{@min} cannot be greater than max #{@max}" if !@min.nil? && !@max.nil? && @min > @max
        end

        def validate_param!(attr_name, params)
          param = params[attr_name]

          return if param.nil? && @allow_nil # when `allow_nil` is `true`, `nil` is allowed

          raise ArgumentError, "parameter #{param} does not support #length" unless param.respond_to?(:length)

          return unless (!@min.nil? && param.length < @min) || (!@max.nil? && param.length > @max)

          raise Grape::Exceptions::Validation.new(params: [@scope.full_name(attr_name)], message: build_message)
        end
        ...
end
```

If you agree, I will be happy to change it.

Comment (dhruvCW) [reply README.md:1716]: I agree but I think the default should be `false`. maybe `ignore_nil` would be a better name ? since we are skipping the validation for `nil` values.

Comment (OuYangJinTing) [reply README.md:1716]: @dhruvCW Thank you for your suggestions.
1. I agree that the default value be to false.
2. `ignore_nil` is a good name, but I use the name `allow_nil`, considering that there is already `allow_blank`, and it should be better to use a similar name.

Comment (dhruvCW) [reply README.md:1716]: >ignore_nil is a good name, but I use the name allow_nil, considering that there is already allow_blank, and it should be better to use a similar name.

makes sense to keep the names consistent 😅 👍

Comment (dblock) [reply README.md:1716]: Yes, the other reason not to change these is to remain positive vs. negative.

**Review by dblock — CHANGES_REQUESTED** [review]
I’ll wait to merge this till everyone above is happy, including myself :) Please say “I’m good with merging this” when you are! And thank you so much for having an interesting conversation!

**Comment by OuYangJinTing** [comment]
@dhruvCW @dblock Hello.
According to the above discussion, I refactored the code once. Please help review it when you have time.

**Comment by dblock** [comment]
How does this interact with `allow_blank`?

Document the new validator in https://github.com/ruby-grape/grape?tab=readme-ov-file#built-in-validators.

File: README.md
Comment (dblock) [inline README.md:1718]: For another PR, I think we're being too polite by saying `please` in the documentation. It's just a reference, no need to be nice :)

File: spec/grape/validations/validators/length_spec.rb
Comment (dblock) [inline spec/grape/validations/validators/length_spec.rb:88]: Add a test with an explicit `allow_nil: false`.

**Comment by dblock** [comment]
I'd love some other people's opinion on whether we do want the `allow_nil` option or reuse `allow_blank` for the length validator, maybe @ericproulx?

**Comment by OuYangJinTing** [comment]
> Ok, it's just an option on `length`, I understand. What happens if I `allow_blank: true` on the `length` validator? I assume it throws an error that it's not something known?

I think it's not suitable to add the `allow_blank` option here because there is already an `allow_blank` validator. 
Instead of adding an `allow_blank` option, it's better to use the `allow_blank` validator directly.

File: spec/grape/validations/validators/length_spec.rb
Comment (dblock) [reply spec/grape/validations/validators/length_spec.rb:88]: We still need one test with the default (without any spec of `allow_nil`).

File: lib/grape/validations/validators/length_validator.rb
Comment (dhruvCW) [inline lib/grape/validations/validators/length_validator.rb:23]: I was looking into the code to understand how `allow_blank` interops with other validators and saw that it's an attribute in the [base class](https://github.com/ruby-grape/grape/blob/master/lib/grape/validations/validators/base.rb#L24).

So should we just use this option instead of introducing `allow_nil` ?

Comment (dblock) [reply lib/grape/validations/validators/length_validator.rb:23]: That was my thinking. 🤷‍♂️

Comment (dblock) [reply lib/grape/validations/validators/length_validator.rb:23]: I can help decide: does `allow_nil` actually behave like `allow_blank` in this implementation? Meaning is `nil` vs. `[]` treated the same?

Comment (dhruvCW) [reply lib/grape/validations/validators/length_validator.rb:23]: technically no since `allow_nil` is for only `nil` values while `allow_blank` would mean ignoring both `nil` and `empty` values. Not sure if we want to use `allow_blank` or `allow_nil`. 

I personally do prefer more specific constraints thus `allow_nil` but I am also okay if we decide to ignore `blank` values if the `allow_blank` flag is set to true.

Comment (dblock) [reply lib/grape/validations/validators/length_validator.rb:23]: @OuYangJinTing write specs for these scenarios and let's decide?

Again, thanks so much for being patient. API changes are one-way, we can't easily undo them, so it's worth spending the time.

Comment (dblock) [reply lib/grape/validations/validators/length_validator.rb:23]: @dhruvCW Introducing `allow_nil` here only will make it potentially different from other parameter types. Without looking at specs and current behavior I would say `allow_blank` for arrays/lists should behave like `blank?` and if we wanted a `allow_nil` then we should probably implement it for all other parameter types next to `allow_blank`.

**Comment by ericproulx** [comment]
> I'd love some other people's opinion on whether we do want the `allow_nil` option or reuse `allow_blank` for the length validator, maybe @ericproulx?

Sorry for my late response. Quite busy. I'll take a closer look this week-end

File: lib/grape/validations/validators/length_validator.rb
Comment (OuYangJinTing) [reply lib/grape/validations/validators/length_validator.rb:23]: > @OuYangJinTing write specs for these scenarios and let's decide?
> 
> Again, thanks so much for being patient. API changes are one-way, we can't easily undo them, so it's worth spending the time.

Sorry, I'm a little confused.
Do you need to roll back to commit [171407aeef935ab918335c5ea324ab18a96a19d4](https://github.com/ruby-grape/grape/commit/171407aeef935ab918335c5ea324ab18a96a19d4) and provide a more detailed explanation?

Comment (dblock) [reply lib/grape/validations/validators/length_validator.rb:23]: Let's wait for @ericproulx to help us out with a strong opinion

Comment (OuYangJinTing) [reply lib/grape/validations/validators/length_validator.rb:23]: Hmm, I'll make the changes after everyone finished feedback.

**Comment by ericproulx** [comment]
There's is still something confusing about the `allow_nil` in association with an array. For instance, if we declare the following:
```ruby
params do
  requires :list, type: [Integer], length: { min: 1, allow_nil: true }
end
```

Does it mean that `[nil]` and `nil` are expected to be valid ? Unfortunately, the `coerce_validator` would raise an error since `nil` is not an Integer.

Although, I've noticed something regarding the `allow_blank` validator. If I'm not mistaken, its evaluation depends on where its declared. For instance, the following test will ✅  since it declared before the length validator (I've added the `fail_fast` on purpose to see which error will be triggered)

```ruby
describe 'when allow_blank is declared before length validator' do
  subject { last_response }

  let(:app) do
    Class.new(Grape::API) do
      params do
        requires :list, type: [Integer], allow_blank: false, length: { max: 1 }, fail_fast: true
      end
      get do
      end
    end
  end
  
  before do
    get '/', list: nil
  end

  it { is_expected.to be_bad_request}
end
```

but this one will ❌  and raise the `ArgumentError: parameter  does not support #length`

```ruby
describe 'when allow_blank is declared after length validator' do
  subject { last_response }

  let(:app) do
    Class.new(Grape::API) do
      params do
        requires :list, type: [Integer], length: { max: 1 }, fail_fast: true, allow_blank: false
      end
      get do
      end
    end
  end
  before do
    get '/', list: nil
  end

  it { is_expected.to be_bad_request}
end
```

Based on that behavior, the `length` validator must not raise an ArgumentError when `param` doesn't `respond_to?(:length)` but `return` to let the `allow_blank` catch the case.
       
Also, regarding the `min` and `max` possible values, I feel that both need to be greater than 0 to also let `allow_blank` cover some cases. `allow_blank: false` would not allow the `nil`, empty strings, strings containing only whitespace characters and `empty?` arrays, hashes.  Though, `[nil]` would not be valid since is not `empty?`.

**Comment by OuYangJinTing** [comment]
@ericproulx Thank you for your detailed insights and explanations with examples.👍

Considering the above, in order to avoid bring too much complexity.
I think the length validator should simply skip nil and explain this situation in the document.

**Comment by dhruvCW** [comment]
> @ericproulx Thank you for your detailed insights and explanations with examples.👍
> 
> Considering the above, in order to avoid bring too much complexity. I think the length validator should simply skip nil and explain this situation in the document.
> 
> cc @dhruvCW @dblock

Should we simply skip all values that don't respond to `length` ? I assume the only time this is true is `nil` since the type coercer should either fail or wrap say an integer if that's passed as the parameter.

**Comment by dblock** [comment]
> > @ericproulx Thank you for your detailed insights and explanations with examples.👍
> > Considering the above, in order to avoid bring too much complexity. I think the length validator should simply skip nil and explain this situation in the document.
> > cc @dhruvCW @dblock
> 
> Should we simply skip all values that don't respond to `length` ? I assume the only time this is true is `nil` since the type coercer should either fail or wrap say an integer if that's passed as the parameter.

We should see what that breaks.

I like the idea that `blank` validator works for anything that `respond_to?(:blank?)` and `nil` validator for everything that `respond_to?(:nil?)` which is I suppose everything. It's probably easier to grok it that way.

**Comment by OuYangJinTing** [comment]
@dblock @dhruvCW According to the feedback, I have adjusted the PR again. Please help me check if it is correct when you are free. Thank you.

**Review by ericproulx — APPROVED** [review]

**Review by dhruvCW — APPROVED** [review]
Thanks for hanging in there 🍻  LGTM 🚀

**Review by dblock — CHANGES_REQUESTED** [review]
Thanks!

Is this a breaking change or a bug fix? If it's a breaking change, then we need to major-bump version 😱 . If it's a bug fix, then move the CHANGELOG line to fixes.

Either way please add a section to UPGRADING.md.

**Comment by OuYangJinTing** [comment]
> Thanks!
> 
> Is this a breaking change or a bug fix? If it's a breaking change, then we need to major-bump version 😱 . If it's a bug fix, then move the CHANGELOG line to fixes.
> 
> Either way please add a section to UPGRADING.md.

I think this is a bug fix, because before, if parameters with types that don't support `#length` method, it directly throws an exception. In order to avoid exceptions, developers need to correct the api parameters by themselves, and after merging this PR, this situation can be avoided, this should be a security change.

**Comment by dblock** [comment]
> > Thanks!
> > Is this a breaking change or a bug fix? If it's a breaking change, then we need to major-bump version 😱 . If it's a bug fix, then move the CHANGELOG line to fixes.
> > Either way please add a section to UPGRADING.md.
> 
> I think this is a bug fix, because before, if parameters with types that don't support `#length` method, it directly throws an exception. In order to avoid exceptions, developers need to correct the api parameters by themselves, and after merging this PR, this situation can be avoided, this should be a smooth change.

@ericproulx Could use a second opinion, please?

**Review by dblock — APPROVED** [review]

**Comment by ericproulx** [comment]
> > > Thanks!
> > > Is this a breaking change or a bug fix? If it's a breaking change, then we need to major-bump version 😱 . If it's a bug fix, then move the CHANGELOG line to fixes.
> > > Either way please add a section to UPGRADING.md.
> > 
> > 
> > I think this is a bug fix, because before, if parameters with types that don't support `#length` method, it directly throws an exception. In order to avoid exceptions, developers need to correct the api parameters by themselves, and after merging this PR, this situation can be avoided, this should be a smooth change.
> 
> @ericproulx Could use a second opinion, please?

Feels like a bug fix to me :)

**Comment by dblock** [comment]
> Feels like a bug fix to me :). Nonetheless, I'm not sure about the UPGRADING comments since its a bug fix.

I think better safe than sorry. Merging.

**Pull request merged**
